hbase-common/src/test/java/org/apache/hadoop/hbase/HBaseJupitorExtension.java

```diff
@@ -0,0 +1,216 @@
+ * JUnit5 can only enforce the timeout for each test method.
+ * <p>
+ * Finally, it also forbid System.exit call in tests. TODO: need to find a new way as
+ * SecurityManager has been removed since Java 21.
+ */
+@InterfaceAudience.Private
+public class HBaseJupitorExtension
+  implements InvocationInterceptor, BeforeAllCallback, AfterAllCallback {
+
+  private static final Logger LOG = LoggerFactory.getLogger(HBaseJupitorExtension.class);
+
+  private static final SecurityManager securityManager = new TestSecurityManager();
+
+  private static final ExtensionContext.Namespace NAMESPACE =
+    ExtensionContext.Namespace.create(HBaseJupitorExtension.class);
+
+  private static final Map<String, Duration> TAG_TO_TIMEOUT =
+    ImmutableMap.of(SmallTests.TAG, Duration.ofMinutes(3), MediumTests.TAG, Duration.ofMinutes(6),
+      LargeTests.TAG, Duration.ofMinutes(13), IntegrationTests.TAG, Duration.ZERO);
+
+  private static final String EXECUTOR = "executor";
+
+  private static final String DEADLINE = "deadline";
+
+  public HBaseJupitorExtension() {
+    super();
+  }
+
+  private Duration pickTimeout(ExtensionContext ctx) {
+    Set<String> timeoutTags = TAG_TO_TIMEOUT.keySet();
+    Set<String> timeoutTag = Sets.intersection(timeoutTags, ctx.getTags());
+    if (timeoutTag.isEmpty()) {
+      fail("Test class " + ctx.getDisplayName() + " does not have any of the following scale tags "
+        + timeoutTags);
+    }
+    if (timeoutTag.size() > 1) {
+      fail("Test class " + ctx.getDisplayName() + " has multiple scale tags " + timeoutTag);
+    }
+    return TAG_TO_TIMEOUT.get(Iterables.getOnlyElement(timeoutTag));
+  }
+
+  @Override
+  public void beforeAll(ExtensionContext ctx) throws Exception {
+    // TODO: remove this usage
+    System.setSecurityManager(securityManager);
+    Duration timeout = pickTimeout(ctx);
+    if (timeout.isZero() || timeout.isNegative()) {
+      LOG.info("No timeout for {}", ctx.getDisplayName());
+      // zero means no timeout
+      return;
+    }
+    Instant deadline = Instant.now().plus(timeout);
+    LOG.info("Timeout for {} is {}, it should be finished before {}", ctx.getDisplayName(), timeout,
+      deadline);
+    ExecutorService executor =
+      Executors.newSingleThreadExecutor(new ThreadFactoryBuilder().setDaemon(true)
+        .setNameFormat("HBase-Test-" + ctx.getDisplayName() + "-Main-Thread").build());
+    Store store = ctx.getStore(NAMESPACE);
+    store.put(EXECUTOR, executor);
+    store.put(DEADLINE, deadline);
+  }
+
+  @Override
+  public void afterAll(ExtensionContext ctx) throws Exception {
+    Store store = ctx.getStore(NAMESPACE);
+    ExecutorService executor = store.remove(EXECUTOR, ExecutorService.class);
+    if (executor != null) {
+      executor.shutdownNow();
+    }
+    store.remove(DEADLINE);
+    // reset secutiry manager
+    System.setSecurityManager(null);
+  }
+
+  private <T> T runWithTimeout(Invocation<T> invocation, ExtensionContext ctx) throws Throwable {
```

Comment:
would it make more sense to use 

https://docs.junit.org/current/api/org.junit.jupiter.api/org/junit/jupiter/api/Assertions.html#assertTimeoutPreemptively(java.time.Duration,org.junit.jupiter.api.function.Executable) 

instead ?

Reply:
It will create a thread pool every time when calling this method, which means the beforeAll, beforeEach and test method will be executed in different threads, which will cause trouble if you use therad local or similar things.